File: backend/app/api/routes/posts.py
from typing import Optional

from fastapi import APIRouter, Depends, File, UploadFile, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_optional_user, require_roles
from app.core.database import get_db
from app.core.roles import COMMAND_ROLES, CONTENT_ROLES
from app.core.uploads import IMAGE_EXTENSIONS, save_upload
from app.models.user import User
from app.schemas.post import PostCreate, PostOut, PostReview
from app.services import post_service

router = APIRouter(prefix="/posts", tags=["posts"])


@router.post(
    "",
    response_model=PostOut,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_roles(*CONTENT_ROLES))],
)
def create_post(
    post_in: PostCreate,
    as_draft: bool = False,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return post_service.create_post(db, post_in, current_user, as_draft=as_draft)


@router.get("", response_model=list[PostOut], status_code=status.HTTP_200_OK)
def list_posts(
    skip: int = 0,
    limit: int = 100,
    category: Optional[str] = None,
    classification: Optional[str] = None,
    status_filter: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: Optional[User] = Depends(get_optional_user),
):
    return post_service.list_posts(
        db, current_user, skip, limit, category, status_filter, classification
    )


@router.get("/{post_id}", response_model=PostOut, status_code=status.HTTP_200_OK)
def get_post(
    post_id: int,
    db: Session = Depends(get_db),
    current_user: Optional[User] = Depends(get_optional_user),
):
    return post_service.get_post_or_404(db, post_id, current_user)


@router.put(
    "/{post_id}",
    response_model=PostOut,
    status_code=status.HTTP_200_OK,
    dependencies=[Depends(require_roles(*CONTENT_ROLES))],
)
def update_post(
    post_id: int,
    post_in: PostCreate,
    as_draft: bool = False,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return post_service.update_post(db, post_id, post_in, current_user, as_draft=as_draft)


@router.post(
    "/{post_id}/submit",
    response_model=PostOut,
    status_code=status.HTTP_200_OK,
    dependencies=[Depends(require_roles(*CONTENT_ROLES))],
)
def submit_post(
    post_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return post_service.submit_post(db, post_id, current_user)


@router.post(
    "/{post_id}/review",
    response_model=PostOut,
    status_code=status.HTTP_200_OK,
    dependencies=[Depends(require_roles(*COMMAND_ROLES))],
)
def review_post(
    post_id: int,
    review_in: PostReview,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return post_service.review_post(db, post_id, review_in, current_user)


@router.post(
    "/{post_id}/thumbnail",
    response_model=PostOut,
    status_code=status.HTTP_200_OK,
    dependencies=[Depends(require_roles(*CONTENT_ROLES))],
)
def upload_thumbnail(
    post_id: int,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    saved = save_upload(file, subdir="posts", allowed_ext=IMAGE_EXTENSIONS)
    return post_service.set_thumbnail(db, post_id, saved, current_user)


@router.delete(
    "/{post_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_roles(*CONTENT_ROLES))],
)
def delete_post(post_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    post_service.delete_post(db, post_id, current_user)
    return None

File: frontend/src/api/posts.ts
import { apiClient } from './client'
import type { Classification } from '../types/common'
import type { Post, PostCategory, PostCreate, PostReview } from '../types/post'

export const postsApi = {
  list: (params?: {
    skip?: number
    limit?: number
    category?: PostCategory
    classification?: Classification
    statusFilter?: string
  }) => {
    const query = new URLSearchParams()
    if (params?.skip !== undefined) query.set('skip', String(params.skip))
    if (params?.limit !== undefined) query.set('limit', String(params.limit))
    if (params?.category) query.set('category', params.category)
    if (params?.classification) query.set('classification', params.classification)
    if (params?.statusFilter) query.set('status_filter', params.statusFilter)
    const qs = query.toString()
    return apiClient.get<Post[]>(`/posts${qs ? `?${qs}` : ''}`)
  },
  get: (id: number) => apiClient.get<Post>(`/posts/${id}`),
  create: (post: PostCreate, asDraft = false) =>
    apiClient.post<Post>(`/posts${asDraft ? '?as_draft=true' : ''}`, post),
  update: (id: number, post: PostCreate, asDraft = false) =>
    apiClient.put<Post>(`/posts/${id}${asDraft ? '?as_draft=true' : ''}`, post),
  /** Chuyển bài từ nháp / bị trả lại sang chờ duyệt (hoặc đã duyệt nếu là chỉ huy). */
  submit: (id: number) => apiClient.post<Post>(`/posts/${id}/submit`, {}),
  review: (id: number, payload: PostReview) => apiClient.post<Post>(`/posts/${id}/review`, payload),
  uploadThumbnail: (id: number, file: File) => {
    const fd = new FormData()
    fd.append('file', file)
    return apiClient.postForm<Post>(`/posts/${id}/thumbnail`, fd)
  },
  remove: (id: number) => apiClient.delete<void>(`/posts/${id}`),
}
